### PR TITLE
Optimize tispark performace

### PR DESCRIPTION
**Pull request by zhexuany**
### What problem does this PR solve? 

This introduces a new codec format to improve the performance on TiFlash.

### CODE REVIEW

File: tikv-client/src/main/java/com/pingcap/tikv/expression/ArithmeticBinaryExpression.java [hunk 1]
```diff
@@ -42,6 +42,11 @@ public static ArithmeticBinaryExpression minus(Expression left, Expression right
     return new ArithmeticBinaryExpression(left.dataType, MINUS, left, right);
   }
 
+  public static ArithmeticBinaryExpression multiply(
```
Comment (windtalker) [inline tikv-client/src/main/java/com/pingcap/tikv/expression/ArithmeticBinaryExpression.java]: what about other arithmetic binary expressions, do them need add similar function?

File: core/src/main/scala/org/apache/spark/sql/TiStrategy.scala [hunk 2]
```diff
@@ -96,8 +96,17 @@ case class TiStrategy(getOrCreateTiContext: SparkSession => TiContext)(sparkSess
   private def useStreamingProcess: Boolean =
     sqlConf.getConfString(TiConfigConst.COPROCESS_STREAMING, "false").toLowerCase.toBoolean
 
-  private def isEnableChunk: Boolean =
-    sqlConf.getConfString(TiConfigConst.ENABLE_CHUNK, "true").toLowerCase.toBoolean
+  // streaming only support TypeDefault. Even we enable chunk, we should still
+  // use TypeDefault.
+  private def getCodecFormat(): EncodeType = {
```
[the remark this answers is not in the document]
Comment (zhexuany) [reply core/src/main/scala/org/apache/spark/sql/TiStrategy.scala]: It is related. If the user  wants to use `TypeStreaming`,  we have to use `TypeDefault` even the user enable `Chunk` codec format.

**Comment by birdstorm** [comment]
@zhexuany please resolve comments

**Comment by zhexuany** [comment]
@birdstorm It seems we have concurrent issues. We are investigating it.

**Comment by birdstorm** [comment]
Can you specify what concurrent issue you are encountering?

File: tikv-client/src/main/java/com/pingcap/tikv/expression/ArithmeticBinaryExpression.java (hunk 1 above)
Comment (zhexuany) [reply tikv-client/src/main/java/com/pingcap/tikv/expression/ArithmeticBinaryExpression.java]: For now, we only find multiply and division need this.

File: core/src/main/scala/org/apache/spark/sql/TiStrategy.scala [hunk 3]
```diff
@@ -96,8 +96,20 @@ case class TiStrategy(getOrCreateTiContext: SparkSession => TiContext)(sparkSess
...
+    val codecFormatStr =
+      sqlConf.getConfString(TiConfigConst.CODEC_FORMAT, "chblock").toLowerCase
+    if (useStreamingProcess) {
+      return EncodeType.TypeDefault
+    }
+    codecFormatStr match {
+      case "chunk"   => EncodeType.TypeChunk
+      case "chblock" => EncodeType.TypeCHBlock
```
Comment (windtalker) [inline core/src/main/scala/org/apache/spark/sql/TiStrategy.scala]: If the store type is tikv, we should not use chblock as the encode type

File: tikv-client/src/main/java/com/pingcap/tikv/TiSession.java [hunk 4]
```diff
@@ -162,7 +162,7 @@ public ExecutorService getThreadPoolForTableScan() {
         if (tableScanThreadPool == null) {
           tableScanThreadPool =
               Executors.newFixedThreadPool(
-                  conf.getTableScanConcurrency(),
+                  conf.getTableScanConcurrency() / conf.getPartitionPerSplit(),
```
Comment (windtalker) [inline tikv-client/src/main/java/com/pingcap/tikv/TiSession.java]: could you give some explanations of what problem do you want to solve by this change, and how/why it works?

File: core/src/main/scala/org/apache/spark/sql/TiStrategy.scala (hunk 3 above)
Comment (zhexuany) [reply core/src/main/scala/org/apache/spark/sql/TiStrategy.scala]: TiKV will return data in `Default` codec. We decode data using codec type in resp.

File: tikv-client/src/main/java/com/pingcap/tikv/TiSession.java (hunk 4 above)
Comment (zhexuany) [reply tikv-client/src/main/java/com/pingcap/tikv/TiSession.java]: Okay, I  will add more comments.

**Pull request merged**